### PR TITLE
Fix: approx_percentile_cont_with_weight Panic

### LINKED ISSUE #12716: Panic in scalar function `approx_percentile_cont_with_weight` (SQLancer)
### Describe the bug

See the reproducer in datafusion-cli (compiled from the latest main, commit 02427674)
```
DataFusion CLI v42.0.0
> create table t1(v1 bool);
0 row(s) fetched.
Elapsed 0.065 seconds.

> insert into t1 values (true);
+-------+
| count |
+-------+
| 1     |
+-------+
1 row(s) fetched.
Elapsed 0.040 seconds.

> select approx_percentile_cont_with_weight('NaN'::Double, 0, 0) from t1 where t1.v1;
thread 'main' panicked at /Users/yongting/Desktop/code/my_datafusion/arrow-datafusion/datafusion/functions-aggregate-common/src/tdigest.rs:647:9:
assertion failed: max.total_cmp(&min).is_ge()
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

### To Reproduce

_No response_

### Expected behavior

It won't cause panic if `from ... where ...` is removed
```
DataFusion CLI v42.0.0
> select approx_percentile_cont_with_weight('NaN'::Double, 0, 0);
+-------------------------------------------------------------------+
| approx_percentile_cont_with_weight(Utf8("NaN"),Int64(0),Int64(0)) |
+-------------------------------------------------------------------+
| inf                                                               |
+-------------------------------------------------------------------+
1 row(s) fetched.
Elapsed 0.099 seconds.
```

### Additional context

Found by SQLancer https://github.com/apache/datafusion/issues/11030

### PR DESCRIPTION
**Pull request by jonathanc-n**
## Which issue does this PR close?

Closes #12716.

## Rationale for this change
Handles case where min/max are infinite or not a number (NaN)

## What changes are included in this PR?
Checks if min/max values are infinite or NaN + tests

## Are these changes tested?
Yes

### CODE REVIEW

File: datafusion/functions-aggregate-common/src/tdigest.rs [hunk 1]
```diff
@@ -641,8 +641,15 @@ impl TDigest {
...
+        let mut min = cast_scalar_f64!(&state[4]);
+
+        if !min.is_finite() {
+            min = f64::NEG_INFINITY;
+        }
+        if !max.is_finite() {
+            max = f64::INFINITY;
+        }
 
         assert!(max.total_cmp(&min).is_ge());
```
Comment (jonahgao) [inline datafusion/functions-aggregate-common/src/tdigest.rs:654]: I think another fix would be to add a guard for this assertion.
```rust
if min.is_finite() && max.is_finite() {
    assert!(max.total_cmp(&min).is_ge());
}
```

The current change causes the following two queries to have different signs. Even though I think their results should be NaN, it's better to keep them consistent.
```sh
> select approx_percentile_cont_with_weight('NaN'::Double, 0, 0);
+-------------------------------------------------------------------+
| approx_percentile_cont_with_weight(Utf8("NaN"),Int64(0),Int64(0)) |
+-------------------------------------------------------------------+
| inf                                                               |
+-------------------------------------------------------------------+
1 row(s) fetched.
Elapsed 0.006 seconds.

> SELECT approx_percentile_cont_with_weight('NaN'::DOUBLE, 0, 0) FROM t1 WHERE t1.v1;
+-------------------------------------------------------------------+
| approx_percentile_cont_with_weight(Utf8("NaN"),Int64(0),Int64(0)) |
+-------------------------------------------------------------------+
| -inf                                                              |
+-------------------------------------------------------------------+
```

Comment (jonathanc-n) [reply datafusion/functions-aggregate-common/src/tdigest.rs:654]: @jonahgao Good catch!

File: datafusion/functions-aggregate-common/src/tdigest.rs [hunk 2]
```diff
@@ -641,10 +641,19 @@ impl TDigest {
...
-        let min = cast_scalar_f64!(&state[4]);
+        let mut max = cast_scalar_f64!(&state[3]);
+        let mut min = cast_scalar_f64!(&state[4]);
 
-        assert!(max.total_cmp(&min).is_ge());
+        if !min.is_finite() {
+            min = f64::NEG_INFINITY;
+        }
+        if !max.is_finite() {
+            max = f64::INFINITY;
```
Comment (jonahgao) [inline datafusion/functions-aggregate-common/src/tdigest.rs:651]: It seems we no longer need these modifications, as they are causing the output to be `-inf`.

File: datafusion/functions-aggregate-common/src/tdigest.rs [hunk 3]
```diff
@@ -641,10 +641,19 @@ impl TDigest {
             v => panic!("invalid centroids type {v:?}"),
         };
 
-        let max = cast_scalar_f64!(&state[3]);
-        let min = cast_scalar_f64!(&state[4]);
+        let mut max = cast_scalar_f64!(&state[3]);
+        let mut min = cast_scalar_f64!(&state[4]);
 
-        assert!(max.total_cmp(&min).is_ge());
+        if !min.is_finite() {
```
Comment (findepi) [inline datafusion/functions-aggregate-common/src/tdigest.rs:647]: Would it be reasonable to filter out NaN values in `merge_sorted_f64` / `merge_unsorted_f64` functions?

Comment (jonathanc-n) [reply datafusion/functions-aggregate-common/src/tdigest.rs]: @findepi Yeah, I think that probably should be the goal here.

**Pull request merged**